[IA-2460] Change R package installation dir

gpcarr opened the pull request:
Tested in a fiab.. created an rstudio runtime using the newest anvil-rstudio-bioconductor image in this PR: https://github.com/anvilproject/anvil-docker/pull/23 and verified that the correct dir is created for R packages and tested that packages install correctly and persist between runtimes

R packages will now be in the directory `/home/rstudio/R/x86_64-pc-linux-gnu-library/{R_VERSION}-${BIOCONDUCTOR_VERSION}` which is more R native package directory. for more questions about this directory, reach out to @nturaga 

The only concern I have is that if a user creates an RStudio VM, installs R packages and then deletes their runtime but keeps their PD then creates a Jupyter runtime they will have the R packages previously installed on their disk in a directory that is bioconductor specific (see screenshot). 
![Screen Shot 2021-01-19 at 3 52 54 PM](https://user-images.githubusercontent.com/23626109/105093934-7b37f080-5a71-11eb-8a08-da370793101a.png)

not sure if this is a blocker, i'm leaning towards leaving it as is because i tested installing the same R package but in jupyter and it works fine and does not even recognize the package in `/home/rstudio/R/x86_64-pc-linux-gnu-library/{R_VERSION}-${BIOCONDUCTOR_VERSION}` (see second screenshot)
![Screen Shot 2021-01-19 at 4 13 27 PM](https://user-images.githubusercontent.com/23626109/105093949-825efe80-5a71-11eb-83b0-1398760c04a5.png)


---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

gpcarr committed:
first pass still have 1 bug to fix

move logic to a script in anvil repo

rm unused var

backwards compatibility for script + packages persisting

nturaga commented on http/src/main/resources/init-resources/gce-init.sh:
```diff
@@ -470,6 +473,23 @@ if [ ! -z "$JUPYTER_DOCKER_IMAGE" ] ; then
+  # for backwards compatibility (Bioconductor images before version 3.12)
+  # Environment variable $BIOCONDUCTOR_VERSION only exists in images >= 3.12 ie us.gcr.io/anvil-gcr-public/anvil-rstudio-bioconductor:0.0.9 introduces this env var
+  if [ -z "$BIOCONDUCTOR_VERSION" ] ; then
+    BIOCONDUCTOR_VERSION=`docker inspect --format '{{.Config.Labels.version}}' ${RSTUDIO_SERVER_NAME} | sed 's/\(^[0-9].[0-9][0-9]\).*/\1/g'`
```
This should work for all the images 0.0.1 - 0.0.9 

```
BIOCONDUCTOR_VERSION=`docker exec rstudio-server R --version | grep 'R version' | sed 's/R version \([0-9].[0-9]\).*/\1/g'`
```

nturaga commented:
The corresponding PR on the anvil-rstudio-bioconductor side of things is https://github.com/anvilproject/anvil-docker/pull/23.

gpcarr commented:
Because we moved the logic to create this directory and give permission to run the script to the anvil-docker repo, these changes are not backwards compatible for older bioconductor images

rtitle commented on http/src/main/resources/init-resources/gce-init.sh:
```diff
@@ -146,9 +147,11 @@ if [ "$IS_GCE_FORMATTED" == "false" ] ; then
-mkdir -p /work/packages
-chmod a+rwx /work/packages
+# Directory to store user installed packages so they persist for non RStudio runtimes
+if [ -z "$RSTUDIO_DOCKER_IMAGE" ] ; then
```
Wonder if this block could be moved to the jupyter-specific setup below

gpcarr replied:
i like that approach better, will update

rtitle commented on http/src/main/resources/init-resources/gce-init.sh:
```diff
@@ -470,6 +473,11 @@ if [ ! -z "$JUPYTER_DOCKER_IMAGE" ] ; then
 
+# RStudio specific setup; only do if RStudio is installed
+if [ ! -z "$RSTUDIO_DOCKER_IMAGE" ] ; then
+  retry 3 docker exec ${RSTUDIO_SERVER_NAME} ${RSTUDIO_SCRIPTS}/set_up_package_dir.sh
```
This might not be backwards compatible -- e.g. this script won't exist on older RStudio images which would cause the init script to fail. Could we check if the script exists in the image somehow? Alternatively we've used this pattern for things that might fail:
```
EXIT_CODE=0
docker_cmd_that_might_fail || EXIT_CODE=$?
if [ $EXIT_CODE -ne 0 ]; then
  echo "something failed"
fi
```

rtitle commented:
A few comments above.. also can you update the PR title, my eyes skip over it :)

gpcarr commented:
@rtitle I added the logic so that these changes are backwards compatible. I added the `EXIT_CODE` so that if the script does not exist for older images, the init script won't fail. I also added [logic](https://github.com/DataBiosphere/leonardo/pull/1821/files#diff-0ede0263691f9df737e1beecb48f32291556d43f643ecd76f4ebcd14314b6299R482) to create the `/home/rstudio/packages` directory if the script fails so that user installed packages for older images still persist. older images will have `R_LIBS='/home/rstudio/packages` in `/usr/local/lib/R/etc/Renviron.site` so we need to make sure this directory gets created for older images. see https://github.com/anvilproject/anvil-docker/blob/master/anvil-rstudio-bioconductor/Dockerfile#L35 for reference. this PR is ready for final review :) i also retested this in a fiab using jupyter images, older rstudio images and the newest rstudio image which is in my anvil-docker PR

rtitle commented on http/src/main/resources/init-resources/gce-init.sh:
```diff
@@ -470,6 +473,17 @@ if [ ! -z "$JUPYTER_DOCKER_IMAGE" ] ; then
+  if [ $EXIT_CODE -ne 0 ]; then
+    echo "RStudio user package installation directory creation failed, creating /packages directory"
+    docker exec ${RSTUDIO_SERVER_NAME} /bin/bash -c "mkdir -p ${RSTUDIO_USER_HOME}/packages && chmod a+rwx ${RSTUDIO_USER_HOME}/packages"
+    exit $EXIT_CODE
```
We probably don't want to exit here, right? E.g. the script should continue on.

gpcarr replied:
ohh yeah probably not - i added it following the other patterns in the script but this is a different case. I'll remove this but quick question - i tested with an older bioconductor image and the script did not fail. i'm guessing `exit` does not mean the script fails but rather just terminates sooner?

rtitle replied:
Yeah looks like this is near the end anyway (only thing left is `docker prune`) so the image probably started up anyway… still better to not `exit` though. :)

gpcarr committed:
dont exit when anvil script fails

rtitle reviewed (approved):
:shipit:

nturaga commented:
Looks good to me too!
